tests/h/views/api/helpers/links_test.py

```diff
@@ -0,0 +1,74 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import pytest
+
+from h.views.api.helpers import links
+
+
+class TestRegisterLink(object):
+    def test_it_creates_attrs_on_registry_if_not_present(
+        self, service_link, versions, pyramid_config
+    ):
+        links.register_link(service_link(), versions, pyramid_config.registry)
```

Comment:
I would have liked to mock the `registry` here, but the way we're messing with attributes on it (see my previous comment) makes this challenging. If I `autospec`, it will break because the autospec'd mock doesn't "know" about the attribute (`api_links`) that gets added. I futzed around for a while to see if I could find a workaround but didn't want to waste too much time on it.

Reply:
Why do you think it would be useful to mock the registry here? I don't really see the advantage to that in this context.

Reply:
Similar to how we use Pyramid's `DummyRequest` instead of mocking the request, I think it's probably better to use the "real" test config rather than mock it anyway. It's based on Pyramid's `testConfig` thing. See <https://docs.pylonsproject.org/projects/pyramid/en/latest/api/testing.html#pyramid.testing.testConfig> and <https://github.com/hypothesis/h/blob/023f1a8efccea4bc1ee6f616c1c509bcbe152df9/tests/h/conftest.py#L220-L230>

It's part of the design of Pyramid that random parts of the code can add things to globals like the request and the config. For example Pyramid extensions work like this, e.g. `pyramid_services` adds `request.find_service()`.

Within an app you can make the different parts of an app sort of "orthogonal" and self-contained because if the `h/foobar/` package contains all the foobar code and also does the work of adding the `request.foobar` property. So all the foobar stuff is contained within one place.

That said, we don't really organise our application code like this. We organise it into `h/models/`, `h/views/`, `h/services/`, etc. So a given foobar feature is likely to be spread around in `h/models/foobar.py`, `h/views/foobar.py`, `h/services/foobar.py`.

From that point of view it might be sensible to have one central place (like `h/request/`?) where all our code that adds stuff on to the request and registry lives?